File: src/main/java/io/github/binout/soccer/infrastructure/mail/MailService.java
package io.github.binout.soccer.infrastructure.mail;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

public interface MailService {

    void sendMail(Mail mail);

    class Mail {
        private final String from;
        private final String subject;
        private final String content;
        private final List<String> tos;

        public Mail(String from, String subject, String content) {
            this.from = from;
            this.subject = subject;
            this.content = content;
            this.tos = new ArrayList<>();
        }

        public Mail addRecipient(String to) {
            this.tos.add(to);
            return this;
        }

        public boolean hasRecipients() {
            return !tos.isEmpty();
        }

        public String from() {
            return from;
        }

        public String subject() {
            return subject;
        }

        public String content() {
            return content;
        }

        public Stream<String> recipients() {
            return tos.stream();
        }
    }
}

File: src/main/kotlin/io/github/binout/soccer/domain/notification/MailService.kt
/*
 * Copyright 2016 Benoît Prioux
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.binout.soccer.domain.notification

import java.util.ArrayList
import java.util.stream.Stream

interface MailService {

    fun sendMail(mail: Mail)

    class Mail(private val from: String, private val subject: String, private val content: String) {
        private val tos: MutableList<String>

        init {
            this.tos = ArrayList()
        }

        fun addRecipient(to: String): Mail {
            this.tos.add(to)
            return this
        }

        fun hasRecipients(): Boolean {
            return !tos.isEmpty()
        }

        fun from(): String {
            return from
        }

        fun subject(): String {
            return subject
        }

        fun content(): String {
            return content
        }

        fun recipients(): Stream<String> {
            return tos.stream()
        }
    }
}
